<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Mi Wang &amp; Douglas Scofield" />

<meta name="date" content="2017-10-02" />

<title>How to use RPASE</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">How to use RPASE</h1>
<h4 class="author"><em>Mi Wang &amp; Douglas Scofield</em></h4>
<h4 class="date"><em>2017-10-02</em></h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The package RPASE (Read-backed Phasing-based ASE detection) implements an algorithm for identifying genes which show allele-specific expression (ASE) on a per-individual and per-gene basis, using RNA-seq read count data. It takes a VCF containing variants phased using read-backed phasing from <a href="https://software.broadinstitute.org/gatk/">Genome Analysis Toolkit</a> and applies an extra-binomial exact test which takes the interdependency of local read counts into consideration. RPASE leverages information across SNPs and controls for overdispersion in read counts. The details of the statistical approach implemented by RPASE can be found in the accompanying paper.</p>
<p>Note that RPASE can be applied to any study unit (genes, exons, etc.) as specified by the annotation supplied.</p>
</div>
<div id="caveat-emptor" class="section level2">
<h2>Caveat Emptor</h2>
<ul>
<li><p>RPASE operates on read counts from phased heterozygous SNPs produced by <a href="https://software.broadinstitute.org/gatk/">GATK</a>. It is up to the user to call, phase and filter SNPs using GATK tools. Also, any known artifacts that could result in spurious ASE calls should be removed prior to analysis.</p></li>
<li><p>RPASE provides a <a href="#estimating-overdispersion">function to estimate overdispersion in read counts</a> in user data. Alternatively, users may provide their own estimate of overdispersion. It is up to the user to decide whether the overdispersion estimate provided by <code>phi.estimate</code> makes biological sense for their data.</p></li>
<li><p>The p-values returned by RPASE are not adjusted for multiple test correction.</p></li>
<li><p>Finally, large variation in coverage of alleles between SNPs within a phased block (e.g., 5|1000 for one SNP and 1000|10 for another) can result in large amounts of memory being used while calculating the exact p-value. Such variation is rare in our experience (less than 0.05% of phased blocks) and may occur due to mapping error, repetitive regions, or inaccurate position annotation. among other causes. The data may be filtered to remove such cases if they occur in the data and memory is limiting.</p></li>
</ul>
</div>
<div id="input-data" class="section level2">
<h2>Input data</h2>
<p>RPASE requires information on (1) genomic variants phased using read-backed phasing in <a href="https://software.broadinstitute.org/gatk/">GATK</a>, and (2) an annotation file describing study units. Genomic variants would typically be provided in a single-sample VCF file, which can be read into R using a variety of packages. For example, with <a href="https://cran.r-project.org/package=vcfR">vcfR</a>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;vcfR&quot;</span>)
the.vcf &lt;-<span class="st"> </span><span class="kw">read.vcfR</span>(<span class="st">&quot;file.vcf&quot;</span>)
phased.variants &lt;-<span class="st"> </span><span class="kw">cbind</span>(the.vcf@fix, the.vcf@gt)</code></pre></div>
<p>The annotation information is contained in a four-column dataframe. This can be read directly from a <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> file which includes interval names, or could be modified from a <a href="http://www.ensembl.org/info/website/upload/gff.html">GFF/GTF</a> file read using, for example, the <a href="https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a> package, which is part of <a href="https://bioconductor.org">BioConductor</a>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicFeatures)
txdb &lt;-<span class="st"> </span><span class="kw">makeTxDbFromGFF</span>(<span class="st">&quot;file.gff&quot;</span>, <span class="dt">format=</span><span class="st">&quot;gff&quot;</span>)
annotation &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">genes</span>(txdb))[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">6</span>)]
<span class="kw">colnames</span>(annotation) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chromosome&quot;</span>,<span class="st">&quot;gene.start&quot;</span>,<span class="st">&quot;gene.end&quot;</span>,<span class="st">&quot;gene.name&quot;</span>)</code></pre></div>
<p>We include some example variant and annotation dataframes as part of this package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RPASE)
<span class="kw">data</span>(example_phased_vcf)
example_phased_vcf</code></pre></div>
<pre><code>##    CHROM     POS ID REF         ALT QUAL FILTER
## 1  chr01   21500  .   G A,&lt;NON_REF&gt; 3000   PASS
## 2  chr01   21530  .   G A,&lt;NON_REF&gt; 3000   PASS
## 3  chr01   21560  .   G A,&lt;NON_REF&gt; 3000   PASS
## 4  chr01   24450  .   G A,&lt;NON_REF&gt; 3000   PASS
## 5  chr01   24500  .   G A,&lt;NON_REF&gt; 3000   PASS
## 6  chr01   24550  .   G A,&lt;NON_REF&gt; 3000   PASS
## 7  chr01   10500  .   G A,&lt;NON_REF&gt; 3000   PASS
## 8  chr01   10550  .   G A,&lt;NON_REF&gt; 3000   PASS
## 9  chr03  255500  .   G A,&lt;NON_REF&gt; 3000   PASS
## 10 chr03  255520  .   G A,&lt;NON_REF&gt; 3000   PASS
## 11 chr03  255550  .   G A,&lt;NON_REF&gt; 3000   PASS
## 12 chr03  255570  .   G A,&lt;NON_REF&gt; 3000   PASS
## 13 chr03  255600  .   G A,&lt;NON_REF&gt; 3000   PASS
## 14 chr03  255660  .   G A,&lt;NON_REF&gt; 3000   PASS
## 15 chr03  255700  .   G A,&lt;NON_REF&gt; 3000   PASS
## 16 chr14 1360500  .   G A,&lt;NON_REF&gt; 3000   PASS
## 17 chr14 1360570  .   G A,&lt;NON_REF&gt; 3000   PASS
##                                                                                              INFO
## 1  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 2  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 3  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 4  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 5  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 6  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 7  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 8  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 9  BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 10 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 11 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 12 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 13 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 14 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 15 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 16 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
## 17 BaseQRankSum=;ClippingRankSum=;DP=;ExcessHet=;MLEAC=;MLEAF=;MQRankSum=;RAW_MQ=;ReadPosRankSum=
##                       FORMAT                       SAMPLE
## 1  GT:AD:DP:GQ:PGT:PID:PL:SB 0/1:52,27,0:::0|1::,,,,,:,,,
## 2  GT:AD:DP:GQ:PGT:PID:PL:SB 0/1:48,25,0:::0|1::,,,,,:,,,
## 3  GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:7,17,0:::1|0::,,,,,:,,,
## 4  GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:70,5,0:::0/1::,,,,,:,,,
## 5  GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:62,4,0:::0|1::,,,,,:,,,
## 6  GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:66,3,0:::0|1::,,,,,:,,,
## 7  GT:AD:DP:GQ:PGT:PID:PL:SB 0/1:31,30,0:::0|1::,,,,,:,,,
## 8  GT:AD:DP:GQ:PGT:PID:PL:SB 0/1:34,32,0:::1|0::,,,,,:,,,
## 9  GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:14,5,0:::0|1::,,,,,:,,,
## 10 GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:15,4,0:::0|1::,,,,,:,,,
## 11 GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:16,3,0:::0|1::,,,,,:,,,
## 12 GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:15,3,0:::0|1::,,,,,:,,,
## 13 GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:4,14,0:::1|0::,,,,,:,,,
## 14 GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:3,16,0:::1|0::,,,,,:,,,
## 15 GT:AD:DP:GQ:PGT:PID:PL:SB  0/1:16,3,0:::0|1::,,,,,:,,,
## 16 GT:AD:DP:GQ:PGT:PID:PL:SB 0/1:29,22,0:::0|1::,,,,,:,,,
## 17 GT:AD:DP:GQ:PGT:PID:PL:SB   0/1:5,4,0:::0|1::,,,,,:,,,</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(example_annotation)
example_annotation</code></pre></div>
<pre><code>##   chromosome gene.start gene.end gene.name
## 1      chr01      20000    25000     GeneA
## 2      chr01      10000    15000     GeneB
## 3      chr03     255000   260000     GeneC
## 4      chr14    1360000  1380000     GeneD</code></pre>
<div id="generating-a-phased-block-list" class="section level3">
<h3>Generating a phased block list</h3>
<p>The phased block list is a list of dataframes containing information for each study unit. RPASE provides the <code>createPhasedBlockList</code> function for creating a phased block list using variant and annotation information in the formats described above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RPASE)
<span class="kw">data</span>(example_phased_vcf)
<span class="kw">data</span>(example_annotation)
phased_block_list &lt;-<span class="st"> </span><span class="kw">createPhasedBlockList</span>(example_phased_vcf, example_annotation)
phased_block_list</code></pre></div>
<pre><code>## $GeneA.1
##   scaffold position index    gene AD1 AD2
## 3    chr01    21500     1 GeneA.1  52  27
## 4    chr01    21530     1 GeneA.1  48  25
## 5    chr01    21560    -1 GeneA.1  17   7
## 
## $GeneA.2
##   scaffold position index    gene AD1 AD2
## 7    chr01    24500     1 GeneA.2  62   4
## 8    chr01    24550     1 GeneA.2  66   3
## 
## $GeneB.1
##   scaffold position index    gene AD1 AD2
## 1    chr01    10500     1 GeneB.1  31  30
## 2    chr01    10550    -1 GeneB.1  32  34
## 
## $GeneC.1
##    scaffold position index    gene AD1 AD2
## 9     chr03   255500     1 GeneC.1  14   5
## 10    chr03   255520     1 GeneC.1  15   4
## 11    chr03   255550     1 GeneC.1  16   3
## 12    chr03   255570     1 GeneC.1  15   3
## 13    chr03   255600    -1 GeneC.1  14   4
## 14    chr03   255660    -1 GeneC.1  16   3
## 15    chr03   255700     1 GeneC.1  16   3
## 
## $GeneD.1
##    scaffold position index    gene AD1 AD2
## 16    chr14  1360500     1 GeneD.1  29  22</code></pre>
<p>The <code>createPhasedBlockList</code> function provides additional options for filtering phased blocks on various criteria. The <code>min.phased.sites</code> option specified the minimum number of phased SNPs that a phased block must contain. The <code>min.coverage</code> option specifies the minimum number of reads a phased SNP must be covered by to be included in the phased block. The <code>heter</code> option specified the minimum number of reads that must be present for each allele for a SNP to be considered heterozygous. For defaults, see the function documentation.</p>
<p>The annotation information lists the study units requested by the user, but the variants may not contain sufficient phasing information to cover a complete study unit. In such cases, a phased block will cover the extent possible. If the phasing of variants within a study unit contains breaks, then the study unit will be covered by multiple phased blocks. In <code>phased_block_list</code> created above, two phased blocks were created for the single annotation entry named <code>GeneA</code>.</p>
<p>Users could also create their own phased block list. Note that the dataframe for each phased block contains three additional columns describing phase orientation and allele coverage. The <code>index</code> column contains 1 indicating a phase of 0|1, and -1 for a phase of 1|0; within phased blocks, these values may differ from those present in the input variant information so that each phased block has a consistent phase index. The <code>AD1</code> and <code>AD2</code> columns contain read counts for alleles one and two.</p>
</div>
</div>
<div id="estimating-overdispersion" class="section level2">
<h2>Estimating overdispersion</h2>
<p>Overdispersion in read counts can be estimated directly from a phased block list using the <code>phi.estimate</code> function. This is the function used to provide the default overdispersion estimate for the <code>runRPASE</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RPASE)
<span class="kw">data</span>(example_phased_block_list)
<span class="kw">phi.estimate</span>(example_phased_block_list)</code></pre></div>
<pre><code>## [1] 1.28235</code></pre>
<p>There are only five phased blocks in <code>example_phased_block_list</code>, so the overdispersion estimate is quite biased, hence a warning about sample size will be generated if these specific commands are run directly. We provide a more realistic estimate for this dataset (0.3) in the example below.</p>
</div>
<div id="running-rpase" class="section level2">
<h2>Running RPASE</h2>
<p>The <code>runRPASE</code> function tests each phased block sequentially, using a default overdispersion estimate calculated from the input phased block list. The value returned by <code>runRPASE</code> is a vector of named p-values, each representing the result of a RPASE test for a single phased block. The null hypothesis for each test is that there is no allele-specific expression within this phased block.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RPASE)
<span class="kw">data</span>(example_phased_block_list)
<span class="kw">runRPASE</span>(example_phased_block_list, <span class="dt">phi=</span><span class="fl">0.3</span>, <span class="dt">quiet=</span><span class="ot">TRUE</span>)</code></pre></div>
<pre><code>##      GeneA.1      GeneA.2      GeneB.1      GeneC.1      GeneD.1 
## 1.682096e-02 2.008879e-09 4.999050e-01 8.481919e-03 2.831401e-01</code></pre>
</div>
<div id="running-rpase-in-parallel" class="section level2">
<h2>Running RPASE in parallel</h2>
<p>For users who want to use multiple tests in parallel, the <code>RPASE</code> function is available, which tests a single phased block. For <code>RPASE</code>, the overdispersion estimate has no default value and must be precomputed using <code>phi.estimate</code> or another method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(example_phased_block_list)
my.phi &lt;-<span class="st"> </span><span class="fl">0.3</span>
<span class="kw">library</span>(parallel)
my.cluster &lt;-<span class="st"> </span><span class="kw">makeCluster</span>(<span class="dv">4</span>)
res &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">parLapply</span>(my.cluster, example_phased_block_list, RPASE, <span class="dt">phi=</span>my.phi))
<span class="kw">stopCluster</span>(my.cluster)</code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
